torch/distributed/checkpoint/state_dict.py

```diff
@@ -220,7 +223,16 @@ def _verify_state_dict(
         )
 
     if info.handle_optim:
-        if not (optim_state_dict and optim_state_dict[STATE]):
+        valid_osd = False
+        if info.flatten_osd:
+            for key in optim_state_dict.keys():
+                if key.startswith(STATE_PREFIX):
+                    valid_osd = True
+                    break
+        else:
+            valid_osd = optim_state_dict and optim_state_dict[STATE]
+
+        if not valid_osd:
             raise RuntimeError(
                 "The option indicates that model state_dict is required to save, "
```

Comment:
I'm a tad confused about this error message--why would saving the model state_dict require the optim state dict? Do we already have an invariant with users that when they request saving or loading the model, they mean to request both model and optim?

Why not "Attempting to save the optim state dict but found no valid state entries"?

Also...what if the state doesn't exist because the optimizer hasn't stepped yet? I can think of the use case where a user attempts saving model and optim state whenever an error happens during training. If an error is thrown in during the model before any optimizer steps have occurred, the optimizer state will be empty and this error will be thrown. Is that intended behavior? Conversely, how come we don't just save nothing if nothing is found?